### PR TITLE
Migration | Vesting functionality

### PR DESCRIPTION
**Pull request by andreanistico**
https://horizenlabs.atlassian.net/browse/HZN-2268

### CODE REVIEW

File: erc20-migration/contracts/LinearTokenVesting.sol [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    uint256 immutable intervalsToClaim;
+    uint256 intervalsAlreadyClaimed;
+
+    event Claimed(uint256 claimAmount, uint256 timestamp);
+
+    error NothingToClaim();
+    error ClaimCompleted();
+
+    constructor(address _token, address _beneficiary, uint256 _amountForEachClaim, uint256 _startTimestamp, uint256 _timeBetweenClaims, uint256 _intervalsToClaim) {
+        token = _token;
```
Comment (paolocappelletti) [inline erc20-migration/contracts/LinearTokenVesting.sol:22]: we could maybe add some checks on  the input parameters?
_token and _beneficiary not null
_token != _beneficiary (otherwise we have a loop :-))
_timeBetweenClaims and _amountForEachClaim> 0
_startTimestamp >= block.timestamp

File: erc20-migration/contracts/LinearTokenVesting.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        uint256 periodsPassed = (block.timestamp - (startTimestamp + timeBetweenClaims * intervalsAlreadyClaimed)) / timeBetweenClaims;
+        if(periodsPassed == 0) revert NothingToClaim();
+
+        uint256 intervalsToClaimNow = _min(intervalsToClaim - intervalsAlreadyClaimed, periodsPassed); 
+        uint256 amountToClaimNow = intervalsToClaimNow * amountForEachClaim;
+        
+        emit Claimed(amountToClaimNow, block.timestamp);
+        intervalsAlreadyClaimed += intervalsToClaimNow;
+        ERC20(token).transfer(beneficiary, amountToClaimNow);
```
Comment (paolocappelletti) [inline erc20-migration/contracts/LinearTokenVesting.sol:42]: maybe we should add an additional check that the contract has enough funds to be trasfered?
I thinks a corner case could happen if by mistake I have set too much _intervalsToClaim
for example:
- _amountForEachClaim = 10
- _intervalsToClaim = 5
but I send only 46 wei to the contract

File: erc20-migration/test/mytest.js
Comment (paolocappelletti) [inline erc20-migration/test/mytest.js:1]: why deleted?

File: erc20-migration/test/zenclaim_test.js
Comment (paolocappelletti) [inline erc20-migration/test/zenclaim_test.js:1]: why deleted?

**Review by paolocappelletti — CHANGES_REQUESTED** [review]

File: erc20-migration/test/zenclaim_test.js
Comment (andreanistico) [reply erc20-migration/test/zenclaim_test.js:1]: I deleted them to execute my contract's tests faster without executing the other tests - i didn't want to commit it. I restored the files

File: erc20-migration/test/mytest.js
Comment (andreanistico) [reply erc20-migration/test/mytest.js:1]: I deleted them to execute my contract's tests faster without executing the other tests - i didn't want to commit it. I restored the files

File: erc20-migration/contracts/LinearTokenVesting.sol (hunk 2 above)
Comment (andreanistico) [reply erc20-migration/contracts/LinearTokenVesting.sol:42]: In that case the transfer will fail triggering its own Insufficient Balance error, no need to add a redundant check

File: erc20-migration/contracts/LinearTokenVesting.sol [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    function claim() public {
+        if(intervalsAlreadyClaimed == intervalsToClaim) revert ClaimCompleted();
+
+        uint256 periodsPassed = (block.timestamp - (startTimestamp + timeBetweenClaims * intervalsAlreadyClaimed)) / timeBetweenClaims;
+        if(periodsPassed == 0) revert NothingToClaim();
+
+        uint256 intervalsToClaimNow = _min(intervalsToClaim - intervalsAlreadyClaimed, periodsPassed); 
+        uint256 amountToClaimNow = intervalsToClaimNow * amountForEachClaim;
+        
+        emit Claimed(amountToClaimNow, block.timestamp);
```
[the remark this answers is not in the document]
Comment (andreanistico) [reply erc20-migration/contracts/LinearTokenVesting.sol:40]: To avoid reentrancy attacks the interactions with other contracts should be the last thing done in a method. Anyway if the transfer fail the entire transaction fails and the event is not emitted

**Review by paolocappelletti — APPROVED** [review]

File: erc20-migration/contracts/LinearTokenVesting.sol (hunk 1 above)
Comment (i-skrypnyk) [reply erc20-migration/contracts/LinearTokenVesting.sol:31]: Also, should we check that the _token has sufficient balance when deploying the contract?
```
    uint256 totalVestingAmount = _amountForEachClaim * _intervalsToClaim;
    uint256 tokenBalance = ERC20(_token).balanceOf(address(this));
    if(tokenBalance < totalVestingAmount) {
        revert("Insufficient token balance for claiming");
    }
```

File: erc20-migration/contracts/LinearTokenVesting.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        timeBetweenClaims = _timeBetweenClaims;
+        intervalsToClaim = _intervalsToClaim;
+
+        startTimestamp = _startTimestamp;
+    }
+
+    function claim() public {
+        if(intervalsAlreadyClaimed == intervalsToClaim) revert ClaimCompleted();
+
+        uint256 periodsPassed = (block.timestamp - (startTimestamp + timeBetweenClaims * intervalsAlreadyClaimed)) / timeBetweenClaims;
```
Comment (i-skrypnyk) [inline erc20-migration/contracts/LinearTokenVesting.sol:43]: not sure if that is a valid case, but can it happen that a block.timestamp would be for the previous interval?
we can add a simple check here just to be sure:
```
uint256 nextClaimTimestamp = startTimestamp + timeBetweenClaims * intervalsAlreadyClaimed;
if (block.timestamp < nextClaimTimestamp) revert NothingToClaim();

uint256 periodsPassed = (block.timestamp - nextClaimTimestamp) / timeBetweenClaims;
```

Comment (andreanistico) [reply erc20-migration/contracts/LinearTokenVesting.sol:43]: if block.timestamp is less than `startTimestamp + timeBetweenClaims * intervalsAlreadyClaimed` my `periodsPassed` formula would go negative, and this would revert the tx because numbers can't overflow/underflow in solidity from 0.8 version (https://docs.soliditylang.org/en/latest/080-breaking-changes.html)

Comment (i-skrypnyk) [reply erc20-migration/contracts/LinearTokenVesting.sol:43]: Nice! Didn't know of this, thanks for sharing!

File: erc20-migration/contracts/LinearTokenVesting.sol (hunk 1 above)
Comment (andreanistico) [reply erc20-migration/contracts/LinearTokenVesting.sol:31]: no but `address(this)` it is!
